controllers/uploadpackage.lua

```diff
@@ -0,0 +1,108 @@
+local app_helpers = require("lapis.application")
+local validate = require("lapis.validate")
+local capture_errors, assert_error = app_helpers.capture_errors, app_helpers.assert_error
+local lfs = require('lfs')
+local Packages = require("models.packages")
+local Users = require("models.users")
+
+local function isFile(name)
+  if not isDir(name) then
+    return os.rename(name,name) and true or false
+  end
+  return false
+end
+
+local function isFileOrDir(name)
+  if type(name)~="string" then return false end
+  return os.rename(name, name)
+end
+
+local function isDir(name)
+  if type(name)~="string" then return false end
+  local cd = lfs.currentdir()
+  local is, err = lfs.chdir(name)
+  lfs.chdir(cd)
+  return is, err
+end
+
+return {
+  GET = capture_errors(function(self)
+    assert_error(self.session.name, self.i18n("err_not_logged_in"))
+    assert_error(self.session.verified, self.i18n("err_email_not_verified"))
+    return { render = "uploadpackage" }
+  end),
+  POST = capture_errors(function(self)
+    -- must be logged in to upload
+    assert_error(self.session.name, self.i18n("err_not_logged_in"))
+    
+    -- must provide a file, a name, and a version
+    validate.assert_valid(self.params, {
+      { "name", exists = true, min_length = 2 },
+      { "version", exists = true },
+      { "file", exists = true, is_file = true }
+    })
+    -- prep some info
+    local fname = self.params.file.filename
+    local fcontent = self.params.file.content
+    local zipheader = "PK"
+    local mudlet_xml_header = "<!DOCTYPE MudletPackage>"
+    local file_extension = self.split_string(fname, "%.")[2]
+    
+    -- validate file extension
+    local proper_extension = file_extension and (file_extension == "mpackage" or file_extension == "xml")
+    assert_error(proper_extension, self.i18n("err_invalid_file_extension"))
+    
+    -- validate the content is proper for the extension claimed
+    if file_extension == "mpackage" then
+      assert_error(string.find(fcontent, zipheader) == 1, self.i18n("err_invalid_mpackage"))
+    elseif file_extension == "xml" then
+      assert_error(string.find(fcontent, mudlet_xml_header), self.i18n("err_invalid_mudlet_xml"))
+    end
+    
+    -- record our current directory so we can change back to it
+    local cwd = lfs.currentdir()
+    
+    -- start in the data directory
+    assert_error(lfs.chdir(self.config.data_dir))
+    
+    -- create username directory if it doesn't exist, validate it's a dir and enter it
+    if not isFileOrDir(self.session.name) then
+      lfs.mkdir(self.session.name)
+    end
+    assert_error(isDir(self.session.name), self.i18n("err_save_file"))
```

Comment:
Will this give us detailed logs to check what went wrong?

Reply:
The nginx logs should contain mention, but I'll double check. I know it spams my console with errors when something goes wrong, so may just be a matter of making sure that goes to a file.